Remove use_c10_dispatcher option

wenleix opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#54969 Remove use_c10_dispatcher option**

With all use cases to hacky wrapper removed, all kernels will be
dispatched with c10 full dispatcher.

Differential Revision: [D27436596](https://our.internmc.facebook.com/intern/diff/D27436596/)

Commit:
Remove use_c10_dispatcher option

With all use cases to hacky wrapper removed, all kernels will be
dispatched with c10 full dispatcher.

Differential Revision: [D27436596](https://our.internmc.facebook.com/intern/diff/D27436596/)

[ghstack-poisoned]

Commit:
add missing file on "Remove use_c10_dispatcher option"


With all use cases to hacky wrapper removed, all kernels will be
dispatched with c10 full dispatcher.

Differential Revision: [D27436596](https://our.internmc.facebook.com/intern/diff/D27436596/)

[ghstack-poisoned]

Commit:
solve merge conflict on "Remove use_c10_dispatcher option"


With all use cases to hacky wrapper removed, all kernels will be
dispatched with c10 full dispatcher.

Differential Revision: [D27436596](https://our.internmc.facebook.com/intern/diff/D27436596/)

[ghstack-poisoned]

Commit:
rebase; fix python lint on "Remove use_c10_dispatcher option"


With all use cases to hacky wrapper removed, all kernels will be
dispatched with c10 full dispatcher.

Differential Revision: [D27436596](https://our.internmc.facebook.com/intern/diff/D27436596/)

[ghstack-poisoned]

wenleix commented on aten/src/ATen/core/op_registration/adaption.h:
```diff
@@ -0,0 +1,61 @@
+namespace impl {
+
+inline c10::optional<MemoryFormat>
+check_tensor_options_and_extract_memory_format(
```
`check_tensor_options_and_extract_memory_format` is still required. But not sure how to rename `hacky_wrapper_for_legacy_signatures.h`. Use `adaption.h` in this PR but let me know if there is any better suggestions. cc @smessmer

ezyang replied:
`check_tensor_options_and_extract_memory_format` is actually a good example of another thing we need to get rid of; it's due to MemoryFormat being a separate argument in some JIT schema but also being specified in TensorOptions.

ezyang reviewed (approved):
wooooo

smessmer commented on tools/codegen/local.py:
```diff
@@ -17,24 +15,10 @@
 
 class Locals(threading.local):
-    use_c10_dispatcher: Optional[UseC10Dispatcher] = None
-_locals = Locals()
```
This looks like we can remove the whole Locals class, cc @ezyang did you want to use it for something else?

wenleix replied:
Discussed this offline with @ezyang . Supporting dynamic scoping during codegen seems a useful feature, plus is not very much code.

smessmer reviewed (approved):
Yay :) Stoked to finally get rid of hacky_wrapper

Commit:
rebase on "Remove use_c10_dispatcher option"


With all use cases to hacky wrapper removed, all kernels will be
dispatched with c10 full dispatcher.

Differential Revision: [D27436596](https://our.internmc.facebook.com/intern/diff/D27436596/)

[ghstack-poisoned]

Commit:
rebase on "Remove use_c10_dispatcher option"


With all use cases to hacky wrapper removed, all kernels will be
dispatched with c10 full dispatcher.

Differential Revision: [D27436596](https://our.internmc.facebook.com/intern/diff/D27436596/)

[ghstack-poisoned]
